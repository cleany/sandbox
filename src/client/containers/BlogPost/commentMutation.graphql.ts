import gql from 'graphql-tag';

export default gql`
  mutation CreateComment($post: ID!, $body: String, $author: String) {
    createComment(data:{ post: $post, body: $body, author: $author })
  }
`;
