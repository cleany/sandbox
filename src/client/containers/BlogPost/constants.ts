export type INPUT_TYPE = 'BLOG_POST/INPUT';
export const INPUT: INPUT_TYPE = 'BLOG_POST/INPUT';
