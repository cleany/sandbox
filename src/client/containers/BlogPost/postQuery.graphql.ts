import gql from 'graphql-tag';

export default gql`
  query Post($id: ID!) {
    post(id: $id) {
      id: _id
      title
      image
      content
      comments {
        body
        author
      }
    }
  }
`;
