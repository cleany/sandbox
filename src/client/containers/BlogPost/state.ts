import { Input } from './types';

export interface State {
  input?: Input;
}

export const initialState: State = {
  input: '',
};

export default State;
