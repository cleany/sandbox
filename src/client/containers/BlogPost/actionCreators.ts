import { ActionCreator, ActionCreatorsMapObject } from 'redux';
import { Input } from './types';
import * as types from './constants';
import {
  InputAction,
} from './actions';

type InputActionCreator = ActionCreator<InputAction>;
export const input: InputActionCreator = (input: Input) => ({
  type: types.INPUT,
  input,
});

export interface ActionCreatorTypes extends ActionCreatorsMapObject {
  input: InputActionCreator;
}
export const actionCreators: ActionCreatorTypes = {
  input,
};

export default actionCreators;
