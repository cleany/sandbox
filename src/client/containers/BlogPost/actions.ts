import { Action } from 'redux';
import { Input } from './types';
import * as T from './constants';

export interface InputAction extends Action {
  type: T.INPUT_TYPE;
  input: Input;
}

export type ActionTypes = InputAction;
