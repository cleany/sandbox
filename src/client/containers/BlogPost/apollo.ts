import { graphql } from 'react-apollo';
import BlogPost from './';
import PostQuery from './postQuery.graphql';
import CommentMutation from './commentMutation.graphql';

export const withData = graphql(PostQuery, {
  skip: ({ params }) => !params || (params && !params.postId),
  options: ({ params }) => ({
    variables: {
      id: params.postId,
    },
  }),
  props: ({ data: { loading, post, error, refetch } }) => ({
    loading,
    post,
    error,
    refetch,
  }),
});

export const withMutation = graphql(CommentMutation, {
  props: ({ mutate }) => ({
    submitComment: mutate,
  }),
});

const withApollo = (component: typeof BlogPost): typeof BlogPost =>
  withData(withMutation(component));

export default withApollo;
