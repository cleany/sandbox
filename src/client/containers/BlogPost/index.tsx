import * as React from 'react';
import { Dispatch, bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { FormControlEventTarget } from '../../types';
import BlogPresentation from './presentation';
import { ThemeColorMap } from '../../types';
import { Props, ActionCreatorTypes, ActionTypes, Input, ErrorType, Post, SubmitComment } from './types';
import { State } from '../../state';
import { selectInput } from './selectors';
import actionCreators from './actionCreators';
import withApollo from './apollo';
import { withTheme } from 'styled-components';

type MapStateToProps = (state: State) => StateProps;
const mapStateToProps: MapStateToProps = (state) => ({
  input: selectInput(state),
});

type MapDispatch = (dispatch: Dispatch<ActionTypes>) => DispatchProps;
const mapDispatchToProps: MapDispatch = (dispatch) => ({
  actions: bindActionCreators(
    actionCreators,
    dispatch,
  ),
});

export interface Props {
  loading: boolean;
  error?: ErrorType;
  post?: Post;
  theme: ThemeColorMap;
  submitComment?: SubmitComment;
  refetch: () => void;
  params: {
    postId: String;
  };
}

export interface StateProps {
  input: Input,
}
export interface DispatchProps {
  actions: ActionCreatorTypes;
}
export type PropTypes = StateProps & DispatchProps & Props;
class BlogPost extends React.Component<PropTypes, undefined> {
  constructor() {
    super();
    this.handleAddingComment = this.handleAddingComment.bind(this);
    this.handleInput = this.handleInput.bind(this);
    this.handleEnterKeyUp = this.handleEnterKeyUp.bind(this);
  }
  private handleInput(e: React.SyntheticEvent<undefined>) {
    e.preventDefault();
    const input = (e.target as FormControlEventTarget).value;
    this.props.actions.input(input);
  }
  private handleAddingComment() {
    const { input } = this.props;
    const author = 'Unknown';
    const { postId } = this.props.params;
    const variables = {
      post: postId as string,
      author,
      body: input,
    };
    this.props.submitComment({ variables }).then(() => {
      this.props.refetch();
      this.props.actions.input('');
    });
  }
  private handleEnterKeyUp(e: React.KeyboardEvent<undefined>) {
    if (e.keyCode === 13) {
      e.preventDefault();
      this.handleAddingComment();
    }
  }
  public render() {
    return (
      <BlogPresentation
        onChange={this.handleInput}
        onSubmit={this.handleAddingComment}
        onKeyUp={this.handleEnterKeyUp}
        {...this.props}
      />
    );
  }
}

const ComponentWithData = withApollo(BlogPost);
const ComponentWithTheme = withTheme(ComponentWithData as typeof BlogPost);

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(ComponentWithTheme);
