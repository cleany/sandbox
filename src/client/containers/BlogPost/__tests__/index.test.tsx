import MockProvider, { getMockStore } from 'redux-mock-provider';
import { shallowToJson } from 'enzyme-to-json';
import { shallow } from 'enzyme';
import * as React from 'react';
import { initialState } from '../state';
import BlogPost from '../';
import BlogPostPresentation from '../presentation';
import props from './__mocks__/blogPostPresentation.mock';
import colors from '../../../theming';

const store = getMockStore({
  key: 'blogPost',
  state: initialState,
});

describe('<BlogPost /> container', () => {
  it('should render as expected', () => {
    const wrapper = shallow(
      <MockProvider store={store}>
        <BlogPost />
      </MockProvider>,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});

describe('<BlogPostPresentation />', () => {
  it('should render as expected', () => {
    const wrapper = shallow(
      <BlogPostPresentation {...props} theme={colors} />,
    );
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
