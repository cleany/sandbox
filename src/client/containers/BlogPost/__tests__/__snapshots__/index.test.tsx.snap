// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BlogPost /> container should render as expected 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(_class) />
</Provider>
`;

exports[`<BlogPostPresentation /> should render as expected 1`] = `
<Section
  alignItems="center"
  backgroundColor="#f5f5f5"
  flexDirection="column"
  full={
    Object {
      "vertical": true,
    }
  }
  pad={
    Object {
      "horizontal": "small",
      "vertical": "large",
    }
  }
>
  <Notification
    message="oops"
    status="error"
  />
  <LoadingIndicator
    isLoading={false}
  />
  <Post
    comment={
      Object {
        "input": "foo",
        "onChange": [Function],
        "onKeyUp": [Function],
        "onSubmit": [Function],
      }
    }
    comments={
      Array [
        Object {
          "author": "Admin User",
          "body": "This is the first comment!",
        },
      ]
    }
    content="Hey there!  Welcome to the blog of Scalable-React-TypeScript! This is just an introductory post, but stay tuned for more!"
    id="58a0ccb112b4c56440f5d6a7"
    image="https://raw.githubusercontent.com/RyanCCollins/cdn/master/stsb-images/ts-resized-2.png"
    title="Welcome!"
  />
</Section>
`;
